packages/consent/consent-tools/src/domain/create-wrapper.ts

```diff
@@ -0,0 +1,185 @@
+import {
+  Categories,
+  CreateWrapper,
+  AnyAnalytics,
+  CDNSettingsIntegrations,
+  CDNSettingsRemotePlugin,
+  InitOptions,
+  CreateWrapperSettings,
+  CDNSettings,
+} from '../types'
+import { validateCategories, validateOptions } from './validation'
+import { createConsentStampingMiddleware } from './consent-stamping'
+import { pipe, pick, uniq } from '../utils'
+import { AbortLoadError, LoadContext } from './load-cancellation'
+
+export const createWrapper: CreateWrapper = (createWrapperOptions) => {
+  const {
+    disableSegmentInitialization,
+    disableConsentRequirement,
+    getCategories,
+    shouldLoad,
+    integrationCategoryMappings,
+    shouldEnableIntegration,
+  } = createWrapperOptions
+
+  validateOptions(createWrapperOptions)
+
+  return (analytics) => {
+    const ogLoad = analytics.load
+
+    const loadWithConsent: AnyAnalytics['load'] = async (
+      settings,
+      options
+    ): Promise<void> => {
+      // do not load anything -- segment included
+      if (await disableSegmentInitialization?.()) {
+        return
+      }
+
+      const consentRequirementDisabled = await disableConsentRequirement?.()
+      if (consentRequirementDisabled) {
+        // ignore consent -- just call analytics.load as usual
+        return ogLoad.call(analytics, settings, options)
+      }
+
+      // use these categories to disable/enable the appropriate device mode plugins
+      let initialCategories: Categories
+      try {
+        initialCategories =
+          (await shouldLoad?.(new LoadContext())) || (await getCategories())
+      } catch (e: unknown) {
+        // consumer can call ctx.cancel({ loadSegmentNormally: true })
+        // to load Segment but disable consent requirement
+        if (e instanceof AbortLoadError) {
+          if (e.loadSegmentNormally === true) {
+            ogLoad.call(analytics, settings, options)
+          }
+        } else {
+          console.error(e)
+        }
+
+        return
+      }
+
+      validateCategories(initialCategories)
```

Comment:
Good q!

Validation error represents an implementation error that I imagined would only be encountered during development -- as, it would abort the entire load process -- so besides the uncaught promise error, analytics won't load at all. I was thinking this would be enough?

When it comes to error handling broadly -- open q. I think that might fit into my private idea for createWrapper returning an object with an event emitter. I started to do that but abandoned it temporarily because I didn't want to copy the event emitter code and I was a little too lazy to spin it into its own lib (since I want this lib to avoid depending on analytics-core)